### PR TITLE
Emissary namespace patch

### PR DESCRIPTION
**Pull request by jonahjon**
Signed-off-by: jonahjon <jonahjones094@gmail.com>

*Issue #, if available:*

*Description of changes:*

Works without and with the `defaultNamespace` field in the package config.

Without 

Screen Shot 2022-08-16 at 10 44 22 AM

With

Screen Shot 2022-08-16 at 10 50 13 AM

### CODE REVIEW

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch [hunk 1]
```diff
@@ -83,6 +100,56 @@ index b974e5198..bab4360f7 100644
...
 +version: 3.0.0-cbf68a7dd53dda13c681af3e19660a5605050839
+diff --git a/charts/emissary-ingress/ci.in/tests/manifests/backend.yaml b/charts/emissary-ingress/ci.in/tests/manifests/backend.yaml
+index b2d9205df..2a23de707 100644
+--- a/charts/emissary-ingress/ci.in/tests/manifests/backend.yaml
++++ b/charts/emissary-ingress/ci.in/tests/manifests/backend.yaml
+@@ -2,38 +2,38 @@
+ apiVersion: getambassador.io/v1
+ kind: Mapping
+ metadata:
+-  name: quote-backend
```
Comment (lewisdiamond) [inline projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:111]: How does this work?

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch [hunk 2]
```diff
@@ -4086,29 +4153,108 @@ index 000000000..8e5f321d8
 +# Main                                                                         #
 +################################################################################
 \ No newline at end of file
+diff --git a/charts/emissary-ingress/readme.tpl b/charts/emissary-ingress/readme.tpl
```
Comment (lewisdiamond) [inline projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:4156]: We should remove this whole file instead.

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch [hunk 3]
```diff
@@ -4489,9 +4678,36 @@ index c166d4fb0..d3d0261b2 100644
...
++  namespace: {{ .Release.Namespace | default .Values.defaultNamespace | quote }}
+   labels:
+     app.kubernetes.io/name: {{ include "ambassador.name" . }}
+     {{- include "ambassador.labels" . | nindent 4 }}
+@@ -99,7 +99,7 @@ spec:
+       {{- end }}
+       serviceAccountName: {{ include "ambassador.serviceAccountName" . }}
+       {{- if .Values.priorityClassName }}
+-      priorityClassName: {{ .Values.priorityClassName | quote }}
++      priorityClassName: {{ .Values.priorityClassName | .Release.Namespace | default .Values.defaultNamespace | quote }}
```
Comment (lewisdiamond) [inline projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:4707]: Why is this using namespace?

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch [hunk 4]
```diff
@@ -4501,15 +4717,327 @@ index 2d06c8127..f1e043136 100644
            imagePullPolicy: {{ .Values.image.pullPolicy }}
            ports:
              {{- range .Values.service.ports }}
+@@ -190,7 +190,7 @@ spec:
+             {{- end }}
+             - name: AMBASSADOR_NAMESPACE
+               {{- if .Values.namespace.name }}
+-              value: {{ .Values.namespace.name | quote }}
++              value: {{ .Values.namespace.name |.Release.Namespace | default .Values.defaultNamespace | quote }}
```
Comment (lewisdiamond) [inline projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:4725]: If we already have `Values.namespace.name`, why do we need `.Values.defaultNamespace`?

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch [hunk 5]
```diff
@@ -4501,15 +4717,327 @@ index 2d06c8127..f1e043136 100644
...
+               {{ else }}
+               valueFrom:
+                 fieldRef:
+@@ -200,8 +200,8 @@ spec:
+               value: {{ include "ambassador.fullname" . }}-agent-cloud-token
+             {{- if .Values.env }}
+             {{- range $key,$value := .Values.env }}
+-            - name: {{ $key | upper | quote}}
+-              value: {{ $value | quote}}
++            - name: {{ $key | upper | .Release.Namespace | default .Values.defaultNamespace | quote}}
```
Comment (lewisdiamond) [inline projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:4735]: This is incorrect, it removes env variables.

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch [hunk 6]
```diff
@@ -4501,15 +4717,327 @@ index 2d06c8127..f1e043136 100644
...
++  namespace: {{ .Release.Namespace | default .Values.defaultNamespace | quote }}
+ spec:
+   port: 8443
+   protocol: HTTPS
+@@ -14,14 +14,14 @@ spec:
+       from: ALL
+   {{- if hasKey .Values.env "AMBASSADOR_ID" }}
+   ambassador_id:
+-  - {{ .Values.env.AMBASSADOR_ID | quote }}
++  - {{ .Values.env.AMBASSADOR_ID |.Release.Namespace | default .Values.defaultNamespace | quote }}
```
Comment (lewisdiamond) [inline projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:4784]: This shouldn't be modified.

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch [hunk 7]
```diff
@@ -4501,15 +4717,327 @@ index 2d06c8127..f1e043136 100644
...
++  namespace: {{ .Release.Namespace | default .Values.defaultNamespace | quote }}
+ spec:
+   port: 8080
+   protocol: HTTP
+@@ -31,6 +31,6 @@ spec:
+       from: ALL
+   {{- if hasKey .Values.env "AMBASSADOR_ID" }}
+   ambassador_id:
+-  - {{ .Values.env.AMBASSADOR_ID | quote }}
++  - {{ .Values.env.AMBASSADOR_ID |.Release.Namespace | default .Values.defaultNamespace | quote }}
```
Comment (lewisdiamond) [inline projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:4801]: This shouldn't be modified.

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch [hunk 8]
```diff
@@ -4501,15 +4717,327 @@ index 2d06c8127..f1e043136 100644
...
++  namespace: {{ .Release.Namespace | default .Values.defaultNamespace | quote }}
+   labels:
+     app.kubernetes.io/name: {{ include "ambassador.name" . }}
+     {{- include "ambassador.labels" . | nindent 4 }}
+@@ -12,7 +12,7 @@ metadata:
+ spec:
+   {{- if .Values.env }}
+   {{- if hasKey .Values.env "AMBASSADOR_ID" }}
+-  ambassador_id: {{ .Values.env.AMBASSADOR_ID | quote }}
++  ambassador_id: {{ .Values.env.AMBASSADOR_ID |.Release.Namespace | default .Values.defaultNamespace | quote }}
```
Comment (lewisdiamond) [inline projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:4822]: Same.

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch [hunk 9]
```diff
@@ -4501,15 +4717,327 @@ index 2d06c8127..f1e043136 100644
...
+diff --git a/charts/emissary-ingress/templates/namespace.yaml b/charts/emissary-ingress/templates/namespace.yaml
+index 4535c74f2..c3cde61b4 100644
+--- a/charts/emissary-ingress/templates/namespace.yaml
++++ b/charts/emissary-ingress/templates/namespace.yaml
+@@ -4,5 +4,5 @@ kind: Namespace
+ metadata:
+   labels:
+     product: aes
+-  name: {{ include "ambassador.namespace" . }}
++  name: {{ .Release.Namespace | default .Values.defaultNamespace | quote }}
```
Comment (lewisdiamond) [inline projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:4835]: We haven't talked about whether or not we should create namespaces automatically. Metallb doesn't do it, not sure about harbor. We should make it consistent.

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch [hunk 10]
```diff
@@ -4501,15 +4717,327 @@ index 2d06c8127..f1e043136 100644
...
+   name: {{ .Values.resolvers.endpoint.name }}
+-  namespace: {{ include "ambassador.namespace" . }}
++  namespace: {{ .Release.Namespace | default .Values.defaultNamespace | quote }}
+   labels:
+     app.kubernetes.io/name: {{ include "ambassador.name" . }}
+     {{- include "ambassador.labels" . | nindent 4 }}
+ {{- if hasKey .Values.env "AMBASSADOR_ID" }}
+ spec:
+-  ambassador_id: {{ .Values.env.AMBASSADOR_ID | quote }}
++  ambassador_id: {{ .Values.env.AMBASSADOR_ID |.Release.Namespace | default .Values.defaultNamespace | quote }}
```
Comment (lewisdiamond) [inline projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:4919]: Shouldn't change.

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch [hunk 11]
```diff
@@ -4501,15 +4717,327 @@ index 2d06c8127..f1e043136 100644
...
+ metadata:
+   name: {{ .Values.resolvers.consul.name }}
+-  namespace: {{ include "ambassador.namespace" . }}
++  namespace: {{ .Release.Namespace | default .Values.defaultNamespace | quote }}
+   labels:
+     app.kubernetes.io/name: {{ include "ambassador.name" . }}
+     {{- include "ambassador.labels" . | nindent 4 }}
+ spec:
+   {{- if hasKey .Values.env "AMBASSADOR_ID" }}
+-  ambassador_id: {{ .Values.env.AMBASSADOR_ID | quote }}
```
Comment (lewisdiamond) [inline projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:4934]: same

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch [hunk 12]
```diff
@@ -4501,15 +4717,327 @@ index 2d06c8127..f1e043136 100644
...
+@@ -22,10 +22,10 @@ metadata:
+     a8r.io/chat: http://a8r.io/Slack
+     a8r.io/bugs: https://github.com/datawire/ambassador/issues
+     a8r.io/support: https://www.getambassador.io/about-us/support/
+-    a8r.io/dependencies: {{ include "ambassador.fullname" . }}-redis.{{ include "ambassador.namespace" . }}
++    a8r.io/dependencies: {{ include "ambassador.fullname" . }}-redis.{{ .Release.Namespace | default .Values.defaultNamespace | quote|.Release.Namespace | default .Values.defaultNamespace | quote }}
+ {{- if .Values.service.annotations }}
+   {{- range $key, $value := .Values.service.annotations }}
+-    {{ $key }}: {{ $value | quote }}
++    {{ $key }}: {{ $value |.Release.Namespace | default .Values.defaultNamespace | quote }}
```
Comment (lewisdiamond) [inline projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:4961]: no change.

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch [hunk 13]
```diff
@@ -4501,15 +4717,327 @@ index 2d06c8127..f1e043136 100644
...
++  namespace: {{ .Release.Namespace | default .Values.defaultNamespace | quote }}
+   labels:
+     app.kubernetes.io/name: {{ include "ambassador.name" . }}
+     {{- include "ambassador.labels" . | nindent 4 }}
+@@ -22,10 +22,10 @@ metadata:
+     a8r.io/chat: http://a8r.io/Slack
+     a8r.io/bugs: https://github.com/datawire/ambassador/issues
+     a8r.io/support: https://www.getambassador.io/about-us/support/
+-    a8r.io/dependencies: {{ include "ambassador.fullname" . }}-redis.{{ include "ambassador.namespace" . }}
++    a8r.io/dependencies: {{ include "ambassador.fullname" . }}-redis.{{ .Release.Namespace | default .Values.defaultNamespace | quote|.Release.Namespace | default .Values.defaultNamespace | quote }}
```
Comment (lewisdiamond) [inline projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:4957]: This doesn't work.

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch [hunk 14]
```diff
@@ -4501,15 +4717,327 @@ index 2d06c8127..f1e043136 100644
...
++  namespace: {{ .Release.Namespace | default .Values.defaultNamespace | quote }}
+   labels:
+     app.kubernetes.io/name: {{ include "ambassador.name" . }}
+     {{- include "ambassador.labels" . | nindent 4 }}
+@@ -20,10 +20,10 @@ metadata:
+     a8r.io/chat: http://a8r.io/Slack
+     a8r.io/bugs: https://github.com/datawire/ambassador/issues
+     a8r.io/support: https://www.getambassador.io/about-us/support/
+-    a8r.io/dependencies: {{ include "ambassador.fullname" . }}-redis.{{ include "ambassador.namespace" . }}
++    a8r.io/dependencies: {{ include "ambassador.fullname" . }}-redis.{{ .Release.Namespace | default .Values.defaultNamespace | quote }}
```
Comment (lewisdiamond) [inline projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:4983]: Why does this change?

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch [hunk 15]
```diff
@@ -4501,15 +4717,327 @@ index 2d06c8127..f1e043136 100644
...
+ {{- if .Values.service.annotations }}
+   {{- range $key, $value := .Values.service.annotations }}
+-    {{ $key }}: {{ $value | quote }}
```
Comment (lewisdiamond) [inline projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:4986]: No change here.

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch [hunk 16]
```diff
@@ -4501,15 +4717,327 @@ index 2d06c8127..f1e043136 100644
...
+diff --git a/charts/emissary-ingress/templates/servicemonitor.yaml b/charts/emissary-ingress/templates/servicemonitor.yaml
+index b2c8122a1..2360b9031 100644
+--- a/charts/emissary-ingress/templates/servicemonitor.yaml
++++ b/charts/emissary-ingress/templates/servicemonitor.yaml
+@@ -3,7 +3,7 @@ apiVersion: monitoring.coreos.com/v1
+ kind: ServiceMonitor
+ metadata:
+   name: {{ include "ambassador.fullname" . }}
+-  namespace: {{ include "ambassador.namespace" . }}
++  namespace: {{ .Release.Namespace | default .Values.defaultNamespace | quote|.Release.Namespace | default .Values.defaultNamespace | quote }}
```
Comment (lewisdiamond) [inline projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:5013]: This is wrong.

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch [hunk 17]
```diff
@@ -4501,15 +4717,327 @@ index 2d06c8127..f1e043136 100644
...
+ nameOverride: ''
+ # Override the generated release name. Defaults to .Release.Name.
+ fullnameOverride: ''
+-# Override the generated release namespace. Defaults to .Release.Namespace.
+-namespaceOverride: ''
++
++defaultNamespace: emissary-system
  
  # Number of Ambassador replicas
 -replicaCount: 3
```
Comment (lewisdiamond) [inline projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:5043]: Why  no replicaCount? (the diff of a patch file is confusing)

**Review by lewisdiamond — CHANGES_REQUESTED** [review]

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch (hunk 14 above)
Comment (jonahjon) [reply projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:4983]: "ambassador.namespace" gets removed

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch (hunk 4 above)
Comment (junshun) [reply projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:4725]: We should replace Values.namespace.name with .Values.defaultNamespace to match with what package controller expects.

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch (hunk 1 above)
Comment (jonahjon) [reply projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:111]: This is part of the tests, but can be removed for now anyway.

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch (hunk 3 above)
Comment (jonahjon) [reply projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:4707]: fixed

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch (hunk 4 above)
Comment (jonahjon) [reply projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:4725]: Values.namespace.name is only for the commercial product, and used an env variable.

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch (hunk 5 above)
Comment (jonahjon) [reply projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:4735]: good catch

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch (hunk 17 above)
Comment (jonahjon) [reply projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:5325]: Fixed.

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch (hunk 16 above)
Comment (jonahjon) [reply projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch]: Fixed

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch (hunk 15 above)
Comment (jonahjon) [reply projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch]: Fixed

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch (hunk 13 above)
Comment (jonahjon) [reply projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch]: Fixed

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch (hunk 12 above)
Comment (jonahjon) [reply projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch]: Fixed

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch (hunk 11 above)
Comment (jonahjon) [reply projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch]: fixed

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch (hunk 10 above)
Comment (jonahjon) [reply projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch]: fixed

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch (hunk 9 above)
Comment (jonahjon) [reply projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch:5139]: Yea that would be a good conversation

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch (hunk 8 above)
Comment (jonahjon) [reply projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch]: fixed

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch (hunk 7 above)
Comment (jonahjon) [reply projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch]: fixed

File: projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch (hunk 6 above)
Comment (jonahjon) [reply projects/emissary-ingress/emissary/helm/patches/0001-helm-changes.patch]: fixed

**Review by lewisdiamond — APPROVED** [review]

**Pull request merged**